data/crac/crac-impl/src/main/java/com/powsybl/openrao/data/cracimpl/NetworkActionImpl.java

```diff
@@ -7,55 +7,120 @@
+            } else if (elementaryAction instanceof SwitchPair switchPair) {
+                return !network.getSwitch(switchPair.getSwitchToOpen().getId()).isOpen() || network.getSwitch(switchPair.getSwitchToClose().getId()).isOpen();
+            } else if (elementaryAction instanceof TerminalsConnectionAction terminalsConnectionAction) {
+                Identifiable<?> element = network.getIdentifiable(terminalsConnectionAction.getElementId());
+                if (element instanceof Branch<?> branch) {
+                    if (terminalsConnectionAction.isOpen()) {
+                        // Line is considered closed if both terminal are connected
+                        return branch.getTerminal1().isConnected() && branch.getTerminal2().isConnected();
+                    } else {
+                        // Line is already considered opened if one of the terminals is disconnected
+                        return !branch.getTerminal1().isConnected() || !branch.getTerminal2().isConnected();
+                    }
+                } else {
+                    throw new NotImplementedException("TerminalsConnectionAction are only on branches for now");
+                }
+            } else if (elementaryAction instanceof SwitchAction switchAction) {
+                Switch aSwitch = network.getSwitch(switchAction.getSwitchId());
+                return aSwitch.isOpen() != switchAction.isOpen();
+            } else {
+                throw new NotImplementedException();
+            }
+        });
     }
 
     @Override
     public boolean apply(Network network) {
-        if (elementaryActions.stream().anyMatch(elementaryAction -> !elementaryAction.canBeApplied(network))) {
+        if (!canBeApplied(network)) {
             return false;
         } else {
-            elementaryActions.forEach(action -> action.apply(network));
+            elementaryActions.forEach(action -> action.toModification().apply(network, true, ReportNode.NO_OP));
             return true;
         }
     }
 
+    @Override
+    public boolean canBeApplied(Network network) {
+        // TODO: To implement on powsybl-core Action
+        return elementaryActions.stream().allMatch(elementaryAction -> {
+            if (elementaryAction instanceof ShuntCompensatorPositionAction shuntCompensatorPositionAction) {
+                ShuntCompensator shuntCompensator = network.getShuntCompensator(shuntCompensatorPositionAction.getShuntCompensatorId());
+                return shuntCompensatorPositionAction.getSectionCount() <= shuntCompensator.getMaximumSectionCount();
+            } else if (elementaryAction instanceof GeneratorAction || elementaryAction instanceof LoadAction || elementaryAction instanceof DanglingLineAction) {
+                return true;
+            } else if (elementaryAction instanceof PhaseTapChangerTapPositionAction) {
+                // TODO : setpoint out of range ?
```

Comment:
If we can change a bit the behavior of canBeApplied (compared to main) and based on my understanding of your needs, I direclty used the check system of core, you can see the code and the impact on test (real impact only on the pst out or range check):
https://github.com/powsybl/powsybl-open-rao/pull/920/commits/2bae784302fb68c223adb1e92dcc2c47c0cbe2bf

Reply:
The change suggested seems dangerous. You don't really expect canBeApplied to actually apply the action, and if I'm not mistaken with this change it would.

Reply:
Yes indeed, I revert that.

Reply:
@phiedw 
Can the transformer for the PhaseTapChangerTapPositionAction be a three WT or only two WT (here for Rao)?